app/parse.py

```diff
@@ -15,9 +26,166 @@ class Product:
     num_of_reviews: int
 
 
+PRODUCT_FIELDS = [field.name for field in fields(Product)]
+
+_driver: WebDriver | None = None
+
+
+def get_driver() -> WebDriver:
+    return _driver
+
+
+def set_driver(new_driver: WebDriver) -> None:
+    global _driver
+    _driver = new_driver
+
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="[%(levelname)8s]:  %(message)s",
+    handlers=[
+        logging.FileHandler("parser.log"),
+        logging.StreamHandler(sys.stdout),
+    ],
+)
+
+
+def parse_single_product(product_element: WebElement) -> Product:
+    return Product(
+        title=product_element.find_element(
+            By.CSS_SELECTOR, ".title"
+        ).get_attribute("title"),
+        description=product_element.find_element(
+            By.CSS_SELECTOR, ".description"
+        ).text,
+        price=float(product_element.find_element(
+            By.CSS_SELECTOR, ".price"
+        ).text.replace("$", "")),
+        rating=len(product_element.find_elements(
+            By.CSS_SELECTOR, ".ratings .ws-icon-star"
+        )),
+        num_of_reviews=int(product_element.find_element(
+            By.CSS_SELECTOR, ".ratings > p.float-end"
+        ).text.split()[0])
+    )
+
+
+def get_single_page_products(driver: WebDriver) -> list[Product]:
+    product_elements = driver.find_elements(
+        By.CSS_SELECTOR, ".thumbnail"
+    )
+    return [
+        parse_single_product(product)
+        for product in product_elements
+    ]
+
+
+def write_products_to_csv(products: [Product], filename: str) -> None:
```

Comment:
The type hint for the `products` parameter should be `List[Product]` instead of `[Product]`. This is a minor issue but correcting it will improve code readability and consistency with the rest of the type hints.